File: string-to-integer-atoi/string-to-interger-atoi.py
class Solution:
    # @return an integer
    def atoi(self, str):
        negativeNum = False
        startingWhiteSpaces = 0
        integerBuf = ''
        intList = ['0','1','2','3','4','5','6','7','8','9']
        if len(str) == 0:
            return 0
        # count white spaces
        while str[startingWhiteSpaces] == ' ':
            startingWhiteSpaces+=1
        string = str[startingWhiteSpaces:]
        
        # check if all white spaces
        if len(string) == 0:
            return 0
        
        # check if a sign is present
        if string[0] == "+":
            string = string[1:]
        elif string[0] == "-":
            negativeNum = True
            string = string[1:]
        
        # collect as many numbers as possible
        for s in string:
            if s in intList:
                integerBuf += s
            else:
                break
            
        result = 0
        if len(integerBuf)<1:
            return 0
            
        elif negativeNum == True:
            result = int(integerBuf) * -1
        else:
            result = int(integerBuf)
        
        # check overflow    
        if result >2147483647:
            return 2147483647
        if result<-2147483648:
            return -2147483648
            
        return result

File: string-to-integer-atoi/string-to-interger-atoi.java
/*
Implement atoi to convert a string to an integer.

Hint: Carefully consider all possible input cases. If you want a challenge, please do not see below and ask yourself what are the possible input cases.

Notes: It is intended for this problem to be specified vaguely (ie, no given input specs). You are responsible to gather all the input requirements up front.
*/

public class Solution {
    public int myAtoi(String str) {
        int index = 0;
        int sign = 1;
        char ch;
        int digit;
        int total = 0;
        //1. Empty string
        if(str.length() == 0) return 0;
        //2. Remove Spaces
        while(str.charAt(index) == ' ' && index < str.length())
        {
            index ++;
        }
        //3. Handle signs
        if(str.charAt(index) == '+' || str.charAt(index) == '-')
        {
            if(str.charAt(index) == '+')
                sign = 1;
            else
                sign = -1;
            index ++;
        }

        //4. Convert number and avoid overflow
        while(index < str.length())
        {
            ch = str.charAt(index);
            if(ch < '0' || ch > '9') break;
            digit = ch - '0';
            //check if total will be overflow after 10 times and add digit
            if(Integer.MAX_VALUE/10 < total || Integer.MAX_VALUE/10 == total && Integer.MAX_VALUE %10 < digit)
                return sign == 1?Integer.MAX_VALUE:Integer.MIN_VALUE;

            total = 10*total + digit;
            index ++;
        }
        return total * sign;
    }
}
